(#49) DonationDetail (도네이션 상세페이지) UI 구현

tngusmiso opened the pull request:
### Related Issue

<!-- 
Use 'resolve' when you have completed the issue and want to close it. -->
resolve: #49 

### What does this PR do?
- 도네이션 상세 페이지 UI 구현 완료
- 그림자(네온사인) 효과 적용 완료
- CustomView 구현 완료
    - **DetailParticipantsProfileImageView** : 프로필 이미지 1개 보여주는 원형 뷰
    - **DetailParticipantsMoreButtonView** : 참여자가 여러명일 때 보여질 `...` 표시 되어 있는 원형 뷰
    - **DetailParticipantsView** : 프로필 이미지 뷰 2개와 더보기 버튼 뷰 1개 곂쳐진 형태의 뷰 
    - -> 참여자 인원수에 따른 변형은 이후에 구현 예정
- 버튼 클릭 액션 함수 선언

### Why are we doing this?
- 더미 데이터(Strings, Images) 사용하여 UI 구성 확인
- 버튼 터치 했을 때 다음 화면으로 이동할 수 있도록 action 함수 선언

### Screenshots
<img src="https://s3.us-west-2.amazonaws.com/secure.notion-static.com/65977a31-8576-49ed-bb8a-d9137a0f1eee/0515DA50-6E95-4BEB-B1D8-965E4333E595.png?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAT73L2G45O3KS52Y5%2F20210306%2Fus-west-2%2Fs3%2Faws4_request&X-Amz-Date=20210306T075404Z&X-Amz-Expires=86400&X-Amz-Signature=769d6ef11d462ce718c435668bdd3b5eec2311071273ef4b1c06a1c068a3de0c&X-Amz-SignedHeaders=host&response-content-disposition=filename%20%3D%220515DA50-6E95-4BEB-B1D8-965E4333E595.png%22" height="600">

`참여자 더보기 버튼`, `후원하기 버튼` 클릭 시 찍히는 로그

<img src="https://s3.us-west-2.amazonaws.com/secure.notion-static.com/2a8e9cef-7b06-49b8-8d4c-a03409298c59/Untitled.png?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAT73L2G45O3KS52Y5%2F20210306%2Fus-west-2%2Fs3%2Faws4_request&X-Amz-Date=20210306T075857Z&X-Amz-Expires=86400&X-Amz-Signature=fdfb89041425b659734cca9f90ec89cc846a9ed7183e3f9b0d9923a3e66dded0&X-Amz-SignedHeaders=host&response-content-disposition=filename%20%3D%22Untitled.png%22">

tngusmiso committed:
💩 (#49) 임시로 DonationDetailViewController로 이동하는 코드 작성
- Apple 로그인 버튼 클릭 시 DonationDetailViewController로 이동

tngusmiso committed:
🍱 (#49) UIColor Extension 추가

tngusmiso committed:
🔧 (#49) 상세페이지 DonationDetailViewController.swift 추가
- 참여자 리스트를 제외한 UI 작업 완료

tngusmiso committed:
🔧 (#49) 상세페이지 DonationDetailViewController+Layout.swift 추가
- 참여자 리스트를 제외한 UI 작업 완료
- Layout 관련 코드만 extension으로 분리

tngusmiso committed:
🍱 (#49) 프로필 이미지 Asset 추가 - profile.png

tngusmiso committed:
🔧 (#49) DetailParticipantsMoreButtonView.swift 추가
- 도네이션 상세 페이지에서 참여자 프로필을 확인할 수 있는 부분
- 더 많은 참여자 보기 버튼

tngusmiso committed:
🔧 (#49) DetailParticipantsProfileImageView.swift 파일 추가
- 도네이션 상세 페이지에서 참여자 프로필 보여주는 부분
- Default Image는 "profile" 로 지정
- 이후에는 url로 이미지 변경 가능

tngusmiso committed:
🔧 (#49) DetailParticipantsView.swift 추가
- 도네이션 상세페이지에서 참여자 프로필을 여러개 보여주는 부분을 하나로 묶은 뷰

tngusmiso committed:
🚧 (#49) 도네이션 상세페이지에 참여자 프로필 뷰 레이아웃 적용

tngusmiso committed:
🚧 (#49) 도네이션 상세페이지에 버튼 탭 액션 함수 추가

tngusmiso committed:
✨ (#49) shadow 이슈 해결
- progressBarView와, giftLabel에 그림자가 생기지 않는 이슈 해결
- layer.shadowOpacity = 1.0 을 명시해주어야 그림자가 보인다.

tngusmiso committed:
✨ (#49) 도네이션 상세 이미지, 프로필 이미지 변경

tngusmiso committed:
♻️  (#49) 길이가 긴 함수 (setDetailGroupView()) 작은 단위로 분리

JeaSungLEE commented on MobalMobal/MobalMobal/DonationDetail/CustomViews/DetailParticipantsView.swift:
```diff
@@ -0,0 +1,48 @@
+    }
+    
+    // MARK: - Methods
+    private func setConstraints() {
```
setConstraints대신에 updateConstraints에서 하는건 어떨까요??

애플자료들을 봐도 여기서 오토레이아웃 잡는 작업을 많이 하더라구요
https://developer.apple.com/documentation/uikit/uiview/1622512-updateconstraints

tngusmiso replied:
오 좋은 정보 감사합니다 👍 
알아보고 적용하도록 하겠습니다

JeaSungLEE commented on MobalMobal/MobalMobal/DonationDetail/DonationDetailViewController+Layout.swift:
```diff
@@ -0,0 +1,180 @@
+        
+        detailImageView.snp.makeConstraints { make in
+            make.top.leading.trailing.equalToSuperview()
+            make.height.equalTo(229)
```
여기 229로 주는것보단 RATIO로 주는것이 조금더 좋을것 같아요.

아이패드나 가로모드가 된다고 하면 가로에 비해 세로가 너무 좁을수도 있을거 같아요.

tngusmiso replied:
넵 반영하겠습니당

tngusmiso committed:
🎨 (#49) 리뷰 반영
- 상단 이미지뷰의 높이를 고정하지 않고 비율로 수정하였습니다.
- 제약조건을 설정하는 코드는 updateViewConstraints를 재정의하여 사용하였습니다.

tngusmiso committed:
🎨 (#49) DonationDetailViewController+Layout 길어진 코드 분리
- nameGiftGroupView와 detailGroupView를 UIComponent 프로퍼티로 처음부터 생성하도록 변경했습니다.
- 각 뷰들에 대한 제약사항을 정하는 함수를 쪼개서 사용하였습니다.

tngusmiso committed:
🔀 (#49) develop 브랜치 Merge 반영
- UIColor+Additions 와 pbxproj 파일 conflict 수정하였습니다.
- UIView에 extension 된 함수인 addSubviews()를 적용하였습니다.

tngusmiso committed:
🎨 (#49) CustomViews는 updateConstraints() 사용
- UIViews의 레이아웃을 설정하는 코드는 updateConstraints()를 재정의해서 사용합니다.
- 위에서 깜빡하고 반영 못한 리뷰!!

tngusmiso commented:
* 🎨 리뷰 반영하였습니다.
* 🔀 `develop` merge conflict 수정하였습니다.

JeongMinHo commented on MobalMobal/MobalMobal/DonationDetail/DonationDetailViewController+Layout.swift:
```diff
@@ -0,0 +1,181 @@
+    private func setDonationButtonConstraints() {
+        donationButton.snp.makeConstraints { make in
+            make.leading.trailing.bottom.equalToSuperview()
+            make.height.equalTo(85)
```
이 부분도 ratio로 줘야 될 것 같습니다. 이 부분이랑 DetailGroupView 제약 조건 확인 한 번 부탁드려유. 디바이스 작을 때 잘리는 것 같아요!

tngusmiso replied:
그러네영 스크롤뷰로 붙여봐야겠네요!

tngusmiso replied:
@JeongMinHo 
가로모드로 해봤는데 버튼 높이를 비율로 주면 좀 어색한 것 같아요..!

|고정 높이|비율|
|---|---|
|<img src="https://s3.us-west-2.amazonaws.com/secure.notion-static.com/d58c9db5-47e9-4c58-920f-a392622ca2a0/Untitled.png?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAT73L2G45O3KS52Y5%2F20210308%2Fus-west-2%2Fs3%2Faws4_request&X-Amz-Date=20210308T063402Z&X-Amz-Expires=86400&X-Amz-Signature=be3c3b7c2b8bf29f1158c5d9c9b468b2e8ee0295fd8094dce65708806ccf83b8&X-Amz-SignedHeaders=host&response-content-disposition=filename%20%3D%22Untitled.png%22" > | <img src="https://s3.us-west-2.amazonaws.com/secure.notion-static.com/ecd4b062-6219-4d24-8f4b-f2633417462b/Untitled.png?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAT73L2G45O3KS52Y5%2F20210308%2Fus-west-2%2Fs3%2Faws4_request&X-Amz-Date=20210308T063546Z&X-Amz-Expires=86400&X-Amz-Signature=e1c26961ce3033a82c72a9bb20b319617c4968cc62ce481a7cabe853af2c7ca3&X-Amz-SignedHeaders=host&response-content-disposition=filename%20%3D%22Untitled.png%22"> |

어떤 게 나을까요??

JeaSungLEE replied:
전 좌측이요

JeongMinHo replied:
@JeaSungLEE 저희 화면 고정되어 있지 않고 스크롤 뷰로 넣어도 되나요?

JeaSungLEE replied:
네네 괜찮을거 같아요

tngusmiso replied:
넵 확인했슴당! 이미지뷰만 ratio로 적용하고, 버튼은 고정 높이, 화면은 스크롤 뷰로 적용해서 다시 푸시할게요

JeongMinHo replied:
@tngusmiso 👍

tngusmiso committed:
💄 (#53) ScrollView 적용

tngusmiso committed:
💄 (#53) progressBarView shadow 적용 위해서 subviews 재배치
- `detailImageView`가 영역 밖으로 벗어나지 않고 크롭되도록 하기 위해 `clipsToBounds`를 true로 설정했습니다.
- `progressBarView`는 `detailImageView`의 하위뷰이므로, 하단 shadow가 잘리는 현상을 발견했습니다.
- subview 관계를 해제하여, 이미지는 이미지대로 크롭되고, progress의 그림자는 그림자대로 잘 보이도록 변경하였습니다.

tngusmiso committed:
🔀 (#49) develop 브랜치 반영

tngusmiso committed:
🎨 (#49) UIFont의 static 메서드 활용하여 폰트 설정

tngusmiso commented:
* 💄 리뷰 반영하였습니다 : **ScrollView**
* 💄 `progressBarView`에서 아랫부분 shadow가 짤리는 현상 해결했습니다.
* 🔀 `develop` 브랜치 반영하였습니다.
* 🎨 UIFont에 만들어둔 static 함수 사용하여 폰트 설정했습니다.

<img src="https://s3.us-west-2.amazonaws.com/secure.notion-static.com/6e107cfe-f505-42ff-a321-97f901eec9bc/ScreenRecording.gif?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAT73L2G45O3KS52Y5%2F20210312%2Fus-west-2%2Fs3%2Faws4_request&X-Amz-Date=20210312T070400Z&X-Amz-Expires=86400&X-Amz-Signature=4b2155a4d8580150fb307acc8b2d213c369fe9b40348e188eddd6e2d06bc8c48&X-Amz-SignedHeaders=host&response-content-disposition=filename%20%3D%22ScreenRecording.gif%22" width=300>

tngusmiso committed:
Merge branch 'develop' into feature/detail-ui

SongSeoYoung reviewed (approved):
수정 이후 가장 최근에 올리신 코멘트의 스크린샷이 저만 안보이나용..? 아무튼 수고하셨어요👏

tngusmiso committed:
🔀 (#49) develop 브랜치 반영
- 로그인 페이지에서 Firebase 인증 성공 시, 임시로 상세화면으로 넘어가도록 구현했습니다.

tngusmiso commented:
gif를 올렸는데 크기가 커서 깨진 듯 합니다ㅜㅜㅜ
가로모드 관련 스크린샷이었는데, 이제는 사용하지 않게 되어 필요는 없을 것 같습니다!

https://s3.us-west-2.amazonaws.com/secure.notion-static.com/6e107cfe-f505-42ff-a321-97f901eec9bc/ScreenRecording.gif?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAT73L2G45O3KS52Y5%2F20210312%2Fus-west-2%2Fs3%2Faws4_request&X-Amz-Date=20210312T070400Z&X-Amz-Expires=86400&X-Amz-Signature=4b2155a4d8580150fb307acc8b2d213c369fe9b40348e188eddd6e2d06bc8c48&X-Amz-SignedHeaders=host&response-content-disposition=filename%20%3D%22ScreenRecording.gif%22
